libtenzir/include/tenzir/chunk.hpp

```diff
@@ -280,54 +280,96 @@ class chunk final : public caf::ref_counted {
 
   template <class Inspector>
   friend bool load_impl(Inspector& f, chunk_ptr& x) {
-    int64_t size = 0;
-    if (!f.apply(size))
+    x = nullptr;
+    if (not f.template begin_object_t<chunk_ptr>()) {
       return false;
-    if (size == chunk::invalid_size) {
-      x = nullptr;
-      return true;
     }
-    if (size == 0) {
-      x = chunk::make_empty();
-      return true;
+    auto has_bytes = false;
+    if (not f.begin_field("bytes", has_bytes)) {
+      return false;
     }
-    auto buffer = std::make_unique<chunk::value_type[]>(size);
-    const auto data = buffer.get();
-    for (auto i = 0; i < size; ++i)
-      if (!f.apply(buffer[i])) {
-        x = nullptr;
+    if (has_bytes) {
+      chunk::size_type size = 0;
+      if (not f.begin_sequence(size)) {
         return false;
       }
-    // Loading the metadata can fail as it wasn't present before Tenzir v4.4.
-    auto metadata = chunk_metadata{};
-    (void)f.apply(metadata);
-    x = chunk::make(
-      data, size,
-      [buffer = std::move(buffer)]() noexcept {
-        static_cast<void>(buffer);
-      },
-      std::move(metadata));
+      auto buffer = std::make_unique_for_overwrite<std::byte[]>(size); // NOLINT
+      for (auto* it = buffer.get(); it != buffer.get() + size; ++it) {
+        if (not f.value(*it)) {
+          return false;
+        }
+      }
+      if (not f.end_sequence()) {
+        return false;
+      }
+      const auto view = chunk::view_type{buffer.get(), size};
+      x = chunk::make(
+        view,
+        [buffer = std::move(buffer)]() noexcept {
+          static_cast<void>(buffer);
+        },
+        chunk_metadata{});
+    }
+    if (not f.end_field()) {
+      return false;
+    }
+    auto has_metadata = false;
+    if (not f.begin_field("metadata", has_metadata)) {
+      return false;
+    }
+    if (has_metadata) {
+      TENZIR_ASSERT(x, "got chunk without data but with metadata");
```

Comment:
Chunks can be user inputs. I feel asserting is too strong of a reaction to this.

Reply:
48a09c3edf9